### ISSUE TITLE
[Bug] `create_kv_cache_func_.defined` fails while hosting mlc llm server.

### ISSUE THREAD

**Issue by yogeshg** [mlc-ai/mlc-llm#2059]
## 🐛 Bug

I got an issue while trying to host a server on macos using "metal" device and "Llama-2-7b"

```
python -m mlc_llm.serve.server --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --model-lib-path dist/prebuilt_libs/Llama-2-7b-chat-hf/Llama-2-7b-chat-hf-q4f16_1-metal.so --device metal --max-total-seq-length 1000
```
```
[2024-03-30 10:32:28] INFO auto_device.py:76: Found device: metal:0
[2024-03-30 10:32:28] INFO chat_module.py:379: Using model folder: .../mlc-llm/dist/Llama-2-7b-chat-hf-q4f16_1-MLC
[2024-03-30 10:32:28] INFO chat_module.py:380: Using mlc chat config: .../mlc-llm/dist/Llama-2-7b-chat-hf-q4f16_1-MLC/mlc-chat-config.json
[2024-03-30 10:32:28] INFO chat_module.py:529: Using library model: dist/prebuilt_libs/Llama-2-7b-chat-hf/Llama-2-7b-chat-hf-q4f16_1-metal.so
[2024-03-30 10:32:28] INFO chat_module.py:379: Using model folder: .../mlc-llm/dist/Llama-2-7b-chat-hf-q4f16_1-MLC
[2024-03-30 10:32:28] INFO chat_module.py:380: Using mlc chat config: .../mlc-llm/dist/Llama-2-7b-chat-hf-q4f16_1-MLC/mlc-chat-config.json
INFO:     Started server process [10448]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
libc++abi: terminating due to uncaught exception of type tvm::runtime::InternalError: [10:32:28] .../mlc-ai-package-self-runner/_work/package/package/mlc-llm/cpp/serve/function_table.cc:209: InternalError: Check failed: (this->create_kv_cache_func_.defined()) is false: 
Stack trace:


zsh: abort      python -m mlc_llm.serve.server --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC   
```

## To Reproduce

1. Setup conda, and install libraries
1. Get repositories etc using this documentation: https://llm.mlc.ai/docs/
1. Host REST API using this tutorial: https://llm.mlc.ai/docs/deploy/rest.html

CLI command that failed, along with logging provided above.

## Expected behavior
Server should start without any issues.

## Environment

 - Platform: Metal
 - Operating system: Mac OS
 - Device: Metal
 - How you installed MLC-LLM: Conda
 - How you installed TVM-Unity (`pip`, source): 
 - Python version (e.g. 3.10): 3.11
 - GPU driver version (if applicable): --
 - CUDA/cuDNN version (if applicable): --
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models): `python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))" | grep HASH
GIT_COMMIT_HASH: 6d47d37dfe0e8f7bd079859d2aa744531887dacb`
 - Any other relevant information:

## Additional context

I had to add `--max-total-seq-length 1000` to my command, because if I didn't, then a subprocess used to estimate this value was failing due to lack of keys in some json file.

**Comment by MasterJH5574** [comment]
@yogeshg thanks for reporting. The prebuilt library specified via `--model-lib-path` in your command does not have the latest MLC serve support. Could you simply remove this, just running the command below?
```
python -m mlc_llm.serve.server --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal --max-total-seq-length 1000
```

> I had to add --max-total-seq-length 1000 to my command, because if I didn't, then a subprocess used to estimate this value was failing due to lack of keys in some json file.

I'm very curious about this. Do you mind sharing more information about the error message so that we can triage and improve?

**Comment by yogeshg** [comment]
Thanks for the quick response, I just tried using without model-lib-path, but it seems required. Here's the error I get:
```
% python -m mlc_llm.serve.server --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal --max-total-seq-length 1000
usage: __main__.py [-h] --model MODEL --model-lib-path MODEL_LIB_PATH [--device DEVICE]
                   [--max-batch-size MAX_BATCH_SIZE]
                   [--max-total-seq-length MAX_TOTAL_SEQ_LENGTH]
                   [--prefill-chunk-size PREFILL_CHUNK_SIZE] [--enable-tracing]
                   [--host HOST] [--port PORT] [--allow-credentials]
                   [--allowed-origins ALLOWED_ORIGINS] [--allowed-methods ALLOWED_METHODS]
                   [--allowed-headers ALLOWED_HEADERS]

```
Starting a new issue about max-total-seq-length: https://github.com/mlc-ai/mlc-llm/issues/2065

**Comment by MasterJH5574** [comment]
Ah yeah sorry my bad. We updated the serve CLI in #2014. So right now the command should be
```diff
- python -m mlc_llm.serve.server --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal --max-total-seq-length 1000
+ python -m mlc_llm serve --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal --max-total-seq-length 1000
```

**Comment by yogeshg** [comment]
That command also doesn't work:
```
% python -m mlc_llm serve --model dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal --max-total-seq-length 1000
------------------------- Usage -------------------------
usage: MLC LLM Serve CLI [-h] [--device DEVICE] [--model-lib-path MODEL_LIB_PATH] [--max-batch-size MAX_BATCH_SIZE] [--max-total-seq-length MAX_TOTAL_SEQ_LENGTH]
                         [--prefill-chunk-size PREFILL_CHUNK_SIZE] [--enable-tracing] [--host HOST] [--port PORT] [--allow-credentials] [--allow-origins ALLOW_ORIGINS]
                         [--allow-methods ALLOW_METHODS] [--allow-headers ALLOW_HEADERS]
                         model

positional arguments:
  model                 A path to ``mlc-chat-config.json``, or an MLC model directory that contains `mlc-chat-config.json`. It can also be a link to a HF repository pointing to an MLC
                        compiled model. (required)

options:
  -h, --help            show this help message and exit
  --device DEVICE       The device used to deploy the model such as "cuda" or "cuda:0". Will detect from local available GPUs if not specified. (default: "auto")
  --model-lib-path MODEL_LIB_PATH
                        The full path to the model library file to use (e.g. a ``.so`` file). If unspecified, we will use the provided ``model`` to search over possible paths. It the model
                        lib path is not found, it will be compiled in a JIT manner. (default: "None")
  --max-batch-size MAX_BATCH_SIZE
                        The maximum allowed batch size set for the KV cache to concurrently support. (default: "80")
  --max-total-seq-length MAX_TOTAL_SEQ_LENGTH
                        The KV cache total token capacity, i.e., the maximum total number of tokens that the KV cache support. This decides the GPU memory size that the KV cache consumes.
                        If not specified, system will automatically estimate the maximum capacity based on the vRAM size on GPU.
  --prefill-chunk-size PREFILL_CHUNK_SIZE
                        The maximum number of tokens the model passes for prefill each time. It should not exceed the prefill chunk size in model config. If not specified, this defaults to
                        the prefill chunk size in model config.
  --enable-tracing      Enable Chrome Tracing for the server. After enabling, you can send POST request to the "debug/dump_event_trace" entrypoint to get the Chrome Trace. For example,
                        "curl -X POST http://127.0.0.1:8000/debug/dump_event_trace -H "Content-Type: application/json" -d '{"model": "dist/llama"}'"
  --host HOST           host name (default: "127.0.0.1")
  --port PORT           port (default: "8000")
  --allow-credentials   allow credentials
  --allow-origins ALLOW_ORIGINS
                        allowed origins (default: "['*']")
  --allow-methods ALLOW_METHODS
                        allowed methods (default: "['*']")
  --allow-headers ALLOW_HEADERS
                        allowed headers (default: "['*']")
------------------------- Error -------------------------
the following arguments are required: model
```
Error message is odd, because I do specify `model`, but there doesn't seem to be a flag for that.

here's the version of the lib I have based on conda list:
```
% conda list | grep mlc             
# packages in environment at /Users/yo/miniconda3/envs/mlc-llm-dev:
mlc-ai-nightly            0.15.dev184              pypi_0    pypi
mlc-llm-nightly           0.1.dev1053              pypi_0    pypi
```

**Comment by MasterJH5574** [comment]
Apologies again... I responded too fast without checking the details. There is no keyword `--model`. Just
```
python -m mlc_llm serve dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal
```

**Comment by yogeshg** [comment]
wfm, thanks!
Server:
```
% python -m mlc_llm serve dist/Llama-2-7b-chat-hf-q4f16_1-MLC --device metal
...
INFO:     127.0.0.1:58646 - "POST /v1/chat/completions HTTP/1.1" 200 OK
```
Client:
```
% curl -H "Content-Type: application/json" -X POST http://127.0.0.1:8000/v1/chat/completions --data $REQUEST
{"id":"chatcmpl-2f2ee2d14a54450fbf5983ebce8e71b2","choices":[{"finish_reason":"stop","index":0,"message":{"content":"Hello there! *adjusts glasses* It's a pleasure to meet you. How may I assist you today? Do you have any questions or tasks you'd like me to help you with? Please feel free to ask!","role":"assistant","name":null,"tool_calls":null,"tool_call_id":null},"logprobs":null}],"created":1711897438,"model":"dist/Llama-2-7b-chat-hf-q4f16_1-MLC","system_fingerprint":"","object":"chat.completion","usage":{"prompt_tokens":35,"completion_tokens":49,"total_tokens":84}
```
Raised https://github.com/mlc-ai/mlc-llm/pull/2066 to fix docs.

**Comment by MasterJH5574** [comment]
Thank you so much for confirmation and the fix! Could you also confirm that #2065 is gone?
